services.py

```diff
@@ -88,42 +76,80 @@ def get_param_from_config(self, config_path: str, param_name: str):
             self.logger.error(f'Config file not found: {config_path}')
             return None
 
-    def get_lat_lon(self):
-        g = geocoder.ip('me')
-        return g.latlng
+    def formatDegreesMinutes(self, coordinates, digits):
+    
+        parts = coordinates.split(".")
+
+        if (len(parts) != 2):
+            return coordinates
+
+        if (digits > 3 or digits < 2):
+            return coordinates
+        
+        left = parts[0]
+        right = parts[1]
+        degrees = str(left[:digits])
+        minutes = str(right[:3])
 
-    def delete_file(self, path):
-        if os.path.exists(path):
+        return degrees + "." + minutes
+
+    def get_lat_lon(self):
+        try:
+                gps = serial.Serial(SERIAL_PORT, baudrate = 9600, timeout = 0.5)
+                
+                data = gps.readline().decode('utf-8').rstrip()
+                message = data[0:6]
+                # print(message)
+                if (message == "$GNRMC"):
+                    # GPRMC = Рекомендуемые минимальные конкретные данные GPS / транзит
+                    # Чтение фиксированных данных GPS является альтернативным подходом, 
+                    # который также работает
+                    parts = data.split(",")
+                    print(parts)
+                    if parts[2] == 'V':
+                        # V = Предупреждение, скорее всего, спутников нет в поле зрения ...
+                        print ("GPS receiver warning")
+                    else:
+                        # Получить данные о местоположении, которые были переданы с сообщением GPRMC. 
+                        # В этом примере интересуют только долгота и широта. Для других значений можно
+                        # обратиться к http://aprs.gids.nl/nmea/#rmc
+                        longitude = self.formatDegreesMinutes(parts[5], 3)
+                        latitude = self.formatDegreesMinutes(parts[3], 2)
+                        self.logger.info("Device position: lon = " + str(longitude) + ", lat = " + str(latitude))
+                else:
+                    # Обрабатывать другие сообщения NMEA и неподдерживаемые строки
+                    pass
+
+                gps.close()
```

Comment:
This will be leaking memory and what not on every exception (e.g. each time the car goes into a tunnel). You need to either put it into a `finally:` block or open the `gps` value using the `with` clause, if it supports that.

Reply:
I didn't write this code, it was one of the ones that were given to me by the employer (I guess they want it that way?)